### PR TITLE
Add Intel  OpenVINO NNCF Quantization Support

### PR DESCRIPTION
**Pull request by kisaragychihaya**
Add Intel  OpenVINO NNCF Quantization Support .Using int8=True and data=datasets.yaml flags 

### Summary
🤖🔌🗜️

This pull request adds support for INT8 quantization for OpenVINO export and improves device handling for TensorRT export. It modifies `ultralytics/engine/exporter.py` to use the NNCF library and the validation dataset for quantization and to store the quantized model paths.

> _Sing, O Muse, of the valiant pull request_
> _That brought to OpenVINO the gift of speed_
> _By quantizing the weights with NNCF's art_
> _And using the validation set to heed_

### Walkthrough
*  Add support for INT8 quantization for OpenVINO export ([link](https://github.com/ultralytics/ultralytics/pull/4628/files?diff=unified&w=0#diff-b9f4cf4260fabd8ffb31e9a85b332a63d39fb49829ce1d98d980b2ec3549322cL370-R370), [link](https://github.com/ultralytics/ultralytics/pull/4628/files?diff=unified&w=0#diff-b9f4cf4260fabd8ffb31e9a85b332a63d39fb49829ce1d98d980b2ec3549322cL377-R434))
* Remove warning and device assignment for TensorRT export ([link](https://github.com/ultralytics/ultralytics/pull/4628/files?diff=unified&w=0#diff-b9f4cf4260fabd8ffb31e9a85b332a63d39fb49829ce1d98d980b2ec3549322cL162-R162))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancement to OpenVINO export to support INT8 quantization.

### 📊 Key Changes
- Removed device specification for TensorRT 'engine' format, now loading model directly using selected device.
- Improved OpenVINO export to include INT8 quantization option, coupling with YOLO dataset for calibration images.
- Added logic to use NNCF (Neural Network Compression Framework) for model quantization, with special processing for input data.
- Implemented quantization transformations and configuration for ignored scopes in the quantization process.
- Integrated model serialization for the quantized OpenVINO model along with metadata storage.

### 🎯 Purpose & Impact
- 🚀 The new INT8 quantization for OpenVINO export aims to reduce model size and increase inference speed while maintaining accuracy.
- 🎛️ This update provides users with additional deployment options, specifically for performance optimization on supported hardware.
- ✨ Developers and practitioners using Ultralytics models can now leverage enhanced OpenVINO support for edge deployments and resource-constrained environments.

### CODE REVIEW

**Comment by kisaragychihaya** [comment]
> **CLA Assistant Lite bot:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> I have read the CLA Document and I sign the CLA
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request

I have read the CLA Document and I sign the CLA

**Comment by kisaragychihaya** [comment]
I have read the CLA Document and I sign the CLA

**Comment by kisaragychihaya** [comment]
recheck

**Comment by glenn-jocher** [comment]
@kisaragychihaya thank you for the details regarding the specific changes in coverage percentage. Keep in mind that while we always aim for high coverage, the actual percentage can fluctuate depending on the complexity and nature of the added code. The decrease in coverage is likely an effect of the new additions that might not be fully covered in the current test suites.

Rest assured, we understand the importance of maintaining good code coverage. We'll evaluate which parts of these changes require more thorough testing to ensure quality and effectiveness of our software. It's situations like these that help us improve our test coverage strategy and implementations.

This helps us ensure that users of the YOLOv8 repo can rely on the quality and robustness of our codebase. Let's continue creating top-notch models for the YOLO community and for Ultralytics!

**Comment by glenn-jocher** [comment]
@kisaragychihaya thank you for the PR! It seems we can't edit this PR directly since it was submitted from an organization account, so you'll need to make some fixes for us:

- Please review failing pre-commit check and apply fixes recommended there.
- Please review failing GPU tests check, it seems like the line you removed from exporter.py is causing this problem.

File: ultralytics/engine/exporter.py [hunk 1]
```diff
@@ -159,10 +159,7 @@ def __call__(self, model=None):
             raise ValueError(f"Invalid export format='{format}'. Valid formats are {fmts}")
         jit, onnx, xml, engine, coreml, saved_model, pb, tflite, edgetpu, tfjs, paddle, ncnn = flags  # export booleans
 
-        # Device
-        if format == 'engine' and self.args.device is None:
-            LOGGER.warning('WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0')
-            self.args.device = '0'
```
Comment (glenn-jocher) [inline ultralytics/engine/exporter.py:165]: Re-introduce these lines as your removal of them is causing GPU checks to fail.

File: ultralytics/engine/exporter.py [hunk 2]
```diff
@@ -367,15 +364,74 @@ def export_openvino(self, prefix=colorstr('OpenVINO:')):
...
+        fq_ov = str(Path(fq) / self.file.with_suffix('.xml').name)
 
         ov_model = mo.convert_model(f_onnx,
                                     model_name=self.pretty_name,
                                     framework='onnx',
                                     compress_to_fp16=self.args.half)  # export
-
-        # Set RT info
+        if self.args.data and self.args.int8:
+            check_requirements('nncf')
```
Comment (glenn-jocher) [inline ultralytics/engine/exporter.py:377]: Please add a minimum `nncf` version if applicable.

File: ultralytics/engine/exporter.py [hunk 3]
```diff
@@ -367,15 +364,74 @@ def export_openvino(self, prefix=colorstr('OpenVINO:')):
...
+            quantized_ov_model.set_rt_info([255.0], ['model_info', 'scale_values'])
+            quantized_ov_model.set_rt_info(self.args.iou, ['model_info', 'iou_threshold'])
+            quantized_ov_model.set_rt_info([v.replace(' ', '_') for k, v in sorted(self.model.names.items())],
+                                 ['model_info', 'labels'])
+            if self.model.task != 'classify':
+                quantized_ov_model.set_rt_info('fit_to_window_letterbox', ['model_info', 'resize_type'])
+            ov.serialize(quantized_ov_model, fq_ov)
+            yaml_save(Path(fq) / 'metadata.yaml', self.metadata)
+
+            # Set RT info
```
Comment (glenn-jocher) [inline ultralytics/engine/exporter.py:434]: Fix indent on this line (should be 4 spaces instead of 8)
